src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+ result, the source Q# program will have to be authored in such a way that the Q# compiler will translate the
+ conditionals into `__quantum__qis__apply_if*` calls. The tracer will add operations from _both branches_ into the
+ layers it creates to compute the upper bound estimate.
+
+Nested conditionals, conditional measurements and conditional tracked operations will _not_ be supported.
+
+__Caching__ (lower priority): It might be a huge perf win if the Resource Tracer could cache statistics for repeated
+ computations. The Tracer will have an option to cache layering results per quantum module if the boundaries of modules
+ are treated as layering barriers.
+
+## Layering ##
+
+_Definition_: ___Time___ is an integer-valued function on all quantum operations in a program (gates, measurements,
+ qubits allocation/release). For each gate there is start and end times. For each qubit, there are times when the qubit
+ is allocated and released. Start time of a gate cannot be less than allocation time of any of the qubits the gate uses.
+ If two gates or measurements use the same qubit, one of the gates must have start time greater than or equal to the end
+ time of the other.
+
+A sequentially executed quantum program can be assigned a trivial time function, when all quantum operations have
+ duration of 1 and unique start times, ordered to match the flow of the program. Layering compresses the timeline by
+ assuming that some operations might be executed simultaneously while allowing for different operations to have various
+ durations.
+
+_Definition_: Provided a valid _time_ function for the program a ___layer of duration N at time T, denoted as L(T,N),___
+ is a subset of operations in the program such that all of these operations have start time greater or equal _T_ and
+ finish time less than _T + N_. The program is ___layered___ if all gates in it are partitioned into layers, that don't
+ overlap in time. The union of all qubits that are involved in operations of a given layer, will be denoted _Qubits(T,N)_.
+
+A sequential program can be trivially layered such that each layer contains exactly one operation. Notice, that the
+ definition of layer doesn't require the gates to be executed _in parallel_. For example, all gates in a fully sequential
+ program can be also placed into a single layer L(0, infinity). Some gates might be considered to be very cheap and take
+ zero time to execute, those gates can be added to a layer even if they act on the same qubit another gate in this layer
+ is acting on and have to be executed sequentially within the layer.
+
+_Definition_: A ___global barrier___ is any operation that acts on _all_ currently allocated qubits. We'll provide means
+ for the clients to inject global barriers equivalent to an identity operator. This will allow the clients to enforce a
+ particular layering structure (because no later operation can sink below the barrier).

```

Comment:
> can sink below the barrier

What does that mean _to sink below the barrier_? (to sink below an operation that acts on all currently allocated qubits)
Based on the context of the preceding paragraphs I imagine a horizontal time line (like OX axis) where the time stamps increase when moving left to right. The fragment _to sink_ makes an impression of _vertical_ movement _across_ the horizontal time line.
> This will allow the clients to enforce a particular layering structure

By this moment would be nice to understand why one may need to enforce a particular layering structure.
> to inject global barriers equivalent to an identity operator

I assume that with _identity operator_ you mean [this](https://en.wikipedia.org/wiki/Identity_function). If that's the case then you may want to make it a link.
Does the expression "_global barriers equivalent to an identity operator_" mean "operations that acts on all currently allocated qubits such that the qubits stay unchanged"? If yes then you may want to add a clarification, e.g.
> We'll provide means for the clients to inject global barriers equivalent to an identity operator (**i.e. to insert operations that acts on all currently allocated qubits, such that the qubits stay unchanged**).

> A **global barrier** is any operation that acts on _all_ currently allocated qubits.

Why is such an operation called _global barrier_? Is it a widely known term? Why isn't it called _all-qubit operation_? (the expression _all-qubit operation_ seems to me clear during the first-most reading even without a definition)

Reply:
When I wrote this I thought I could define the barrier in terms of quantum operations, but in reality it's not. For example, it won't let through the qubits, that are allocated _after_ the barrier was injected, so the current definition doesn't hold.

Researches want to exercise some control over layering and barriers is one of the means to do so. The question "why they want to do this" is out of scope of this design document.